sdk/storage/Azure.Storage.Blobs/src/BlockBlobWriteStream.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            if (_buffer.Length > 0)
+            {
+                _buffer.Position = 0;
+
+                string blockId = GenerateBlockId(_position + _buffer.Length);
+
+                if (async)
+                {
+                    await _blockBlobClient.StageBlockAsync(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    _blockBlobClient.StageBlock(
+                        base64BlockId: blockId,
+                        content: _buffer,
+                        conditions: _conditions,
+                        progressHandler: _progressHandler,
+                        cancellationToken: cancellationToken);
+                }
+
+                _blockIds.Add(blockId);
+                _buffer.Clear();
+            }
+        }
+
+        protected override async Task FlushInternal(bool async, CancellationToken cancellationToken)
+        {
+            await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+            if (async)
+            {
+                await _blockBlobClient.CommitBlockListAsync(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                _blockBlobClient.CommitBlockList(
+                    base64BlockIds: _blockIds,
+                    conditions: _conditions,
+                    cancellationToken: cancellationToken);
+            }
+        }
+
+        protected override void ValidateBufferSize(long bufferSize)
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
+            }
+
+            if (bufferSize > Constants.Blob.Block.MaxStageBytes)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), $"Must <= {Constants.Blob.Block.MaxStageBytes}");
+            }
+        }
+
+        private static string GenerateBlockId(long offset)
+        {
+            // TODO #8162 - Add in a random GUID so multiple simultaneous
+            // uploads won't stomp on each other and the first to commit wins.
+            // This will require some changes to our test framework's
+            // RecordedClientRequestIdPolicy.
+            byte[] id = new byte[48]; // 48 raw bytes => 64 byte string once Base64 encoded
+            BitConverter.GetBytes(offset).CopyTo(id, 0);
+            return Convert.ToBase64String(id);
+        }
+    }
```

Comment:
One option would be to create `internal IBlockIdGenerator` and substitute with something predictable in tests.

Reply:
This was copied from [BlockBlobClient.PartitionedUploader.GenerateBlockId()](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs#L2018)

Reply:
That'd be another reason to extract this logic from uploader to something in shared source so that both components can use the same.

Reply:
This is already predictable. It was made that way to be producible in recorded tests. An object is probably the best way to share code, but I don't think we need to interface it or anything. Mocks are suitable enough for tests if we ever change this to a non-predictable system (we may need a new system if we support appends based on other discussion).

Reply:
This was copy + pasted from the parallel uploader, including the comment.  I'm fine moving the static method to Common.

Reply:
Moved to common in next iteration.